File: lib/services/m3u_parser.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class M3UParser {
  Future<List<Channel>> parseFromUrl(String url) async {
    final response = await http.get(Uri.parse(url));
    
    if (response.statusCode == 200) {
      return parseContent(response.body);
    } else {
      throw Exception('Failed to load M3U file');
    }
  }
  
  List<Channel> parseContent(String content) {
    final lines = content.split('\n');
    final List<Channel> channels = [];
    
    Channel? currentChannel;
    
    for (var line in lines) {
      line = line.trim();
      
      if (line.startsWith('#EXTINF:')) {
        // Parse channel info
        final channel = _parseExtinf(line);
        if (channel != null) {
          currentChannel = channel;
        }
      } else if (line.isNotEmpty && !line.startsWith('#') && currentChannel != null) {
        // This is the URL for the current channel
        channels.add(Channel(
          name: currentChannel.name,
          url: line,
          group: currentChannel.group,
          logo: currentChannel.logo,
        ));
        currentChannel = null;
      }
    }
    
    return channels;
  }
  
  Channel? _parseExtinf(String line) {
    try {
      // Remove #EXTINF:
      line = line.substring(8);
      
      // Find the comma separating duration and name
      final commaIndex = line.indexOf(',');
      if (commaIndex == -1) return null;
      
      // Get channel name
      final name = line.substring(commaIndex + 1).trim();
      
      // Parse attributes before the comma
      final attributes = line.substring(0, commaIndex);
      
      String? group;
      String? logo;
      
      // Look for group-title
      final groupMatch = RegExp(r'group-title="([^"]+)"').firstMatch(attributes);
      if (groupMatch != null) {
        group = groupMatch.group(1);
      }
      
      // Look for tvg-logo
      final logoMatch = RegExp(r'tvg-logo="([^"]+)"').firstMatch(attributes);
      if (logoMatch != null) {
        logo = logoMatch.group(1);
      }
      
      return Channel(name: name, url: '', group: group, logo: logo);
    } catch (e) {
      return null;
    }
  }
}

File: js/m3u-parser.js
class M3UParser {
    constructor() {
        this.playlist = [];
    }
    
    async parseFromURL(url) {
        try {
            const response = await fetch(url);
            const text = await response.text();
            return this.parse(text);
        } catch (error) {
            throw new Error('Failed to load M3U from URL: ' + error.message);
        }
    }
    
    async parseFromFile(file) {
        return new Promise((resolve, reject) => {
            const reader = new FileReader();
            
            reader.onload = (e) => {
                try {
                    const playlist = this.parse(e.target.result);
                    resolve(playlist);
                } catch (error) {
                    reject(error);
                }
            };
            
            reader.onerror = () => {
                reject(new Error('Failed to read file'));
            };
            
            reader.readAsText(file);
        });
    }
    
    parse(m3uContent) {
        const lines = m3uContent.split('\n');
        const playlist = [];
        let currentEntry = {};
        
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i].trim();
            
            if (line.startsWith('#EXTINF:')) {
                // Nxjerr informacionin nga linja EXTINF
                currentEntry = this.parseExtinf(line);
            } else if (line.startsWith('#EXTGRP:')) {
                // Nxjerr grupin
                currentEntry.group = line.replace('#EXTGRP:', '').trim();
            } else if (line.startsWith('#EXTIMG:')) {
                // Nxjerr imazhin
                currentEntry.logo = line.replace('#EXTIMG:', '').trim();
            } else if (line.startsWith('http')) {
                // Kjo është URL e stream-it
                currentEntry.url = line;
                
                // Shto në playlist vetëm nëse ka URL
                if (currentEntry.url) {
                    playlist.push({...currentEntry});
                }
                
                // Reset për entry-n e radhës
                currentEntry = {};
            }
        }
        
        return playlist;
    }
    
    parseExtinf(extinfLine) {
        const entry = {};
        
        // Nxjerr kohëzgjatjen dhe emrin
        const matches = extinfLine.match(/#EXTINF:(-?\d+)\s*(.*)/);
        if (matches) {
            entry.duration = parseInt(matches[1]);
            const rest = matches[2];
            
            // Nxjerr metadata shtesë
            const tvgIdMatch = rest.match(/tvg-id="([^"]+)"/);
            if (tvgIdMatch) entry.tvgId = tvgIdMatch[1];
            
            const tvgNameMatch = rest.match(/tvg-name="([^"]+)"/);
            if (tvgNameMatch) entry.tvgName = tvgNameMatch[1];
            
            const tvgLogoMatch = rest.match(/tvg-logo="([^"]+)"/);
            if (tvgLogoMatch) entry.logo = tvgLogoMatch[1];
            
            const groupMatch = rest.match(/group-title="([^"]+)"/);
            if (groupMatch) entry.group = groupMatch[1];
            
            // Nxjerr emrin e kanalit (pas presjes)
            const nameMatch = rest.match(/,(.+)$/);
            if (nameMatch) entry.name = nameMatch[1].trim();
        }
        
        return entry;
    }
}
